### PR TITLE
Remove sticky posts setting when we inherit the query

### LINKED ISSUE #38979: Sticky Posts not sticking
### Description

In TT2 (and all block themes I tested) when I set a post to "Sticky" it doesn't seem to have any impact on the order that the posts are displayed in.

I first noticed this when investigating a [related issue](https://github.com/WordPress/gutenberg/pull/38422#issuecomment-1047352619), but it seems to have become more broken since then...

### Step-by-step reproduction instructions

- Switch to a block theme (e.g. TT2)
- Create a sticky post
- Look at a template with a query loop in it (e.g. home.html)
- Confirm that the sticky post isn't at the top.

### Screenshots, screen recording, code snippet

Screenshot 2022-02-22 at 10 53 19
Screenshot 2022-02-22 at 10 53 45

### Environment info

- 5.9, trunk, TT2

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by draganescu**
## What?
Fixes #38979

## Why?
Because the setting not just doesn't work.

## How?
We hide the setting when inherit is true.

## Testing Instructions

1. Switch to a block theme (e.g. TT2)
2. Create a sticky post
3. Look at a template with a query loop in it (e.g. home.html)
4. Confirm that if the block has inherit query you cannot set sticky posts.

## Notes
This PR has explored allowing sticky post behaviour on inherit but the result was consistently unreliable, one or another query var would be overwritten either by WP_Query or by the query block.

Therefore the PR hides the setting and fixes the server rendering of the block to truly inherit, without interfering with the current WP_Query in any way.

Co-authored-by: Jonny Harris <237508+spacedmonkey@users.noreply.github.com>
Co-authored-by: Huub <50170696+huubl@users.noreply.github.com>

### CODE REVIEW

**Review by scruffian — APPROVED** [review]
This fixes it for me in the front end. We still need to fix the editor as well, but that can happen in a followup.

**Comment by carolinan** [comment]
I am not able to reproduce the **bug**,  I don't want to be rude but it would help me if the instructions were more detailed.
The issues says:

- Create a sticky post
- Look at a template with a query loop in it (e.g. home.html)

What settings *exactly* does the query loop need to have for me to reproduce the issue?

>setting the inherit global query setting in the query block

I am assuming this means the option labeled with "Inherit query from template"
But I honestly do not see the bug with or without this setting, with or without the PR, using 6.0 beta or 5.9.3.

**Comment by scruffian** [comment]
For me the reproduction steps are these:

1. Switch to TT2
2. Create an old sticky post
3. Open the home page
4. The sticky post is not at the top of the list

File: packages/block-library/src/post-template/index.php [hunk 1]
```diff
@@ -19,14 +19,22 @@ function render_block_core_post_template( $attributes, $content, $block ) {
...
 	$use_global_query = ( isset( $block->context['query']['inherit'] ) && $block->context['query']['inherit'] );
 	if ( $use_global_query ) {
 		global $wp_query;
 		if ( $wp_query && isset( $wp_query->query_vars ) && is_array( $wp_query->query_vars ) ) {
 			// Unset `offset` because if is set, $wp_query overrides/ignores the paged parameter and breaks pagination.
 			unset( $query_args['offset'] );
-			$query_args = wp_parse_args( $wp_query->query_vars, $query_args );
+
+			// merge the user settings into the defaults of the query
+			$query_args = wp_parse_args( $query_args, $wp_query->query_vars );
```
Comment (ntsekouras) [inline packages/block-library/src/post-template/index.php]: We can't do that because it breaks the whole query. There are properties like `posts_per_page` for example that are set in `query_args` and this way we keep that values and not from the global query.

**Review by ntsekouras — COMMENTED** [review]
Thanks for the PR @draganescu ! Sticky posts have many nuances in the Query Loop context.

For example [your comment](https://github.com/WordPress/gutenberg/issues/38979#issuecomment-1110952341) here:
>Another problem here is that when Inherit query from template is set the sticky preferences should either be hidden or they should be merged in the global query. Which one is more expected for the user? 

I was thinking a while ago to hide them, but it seems preferable if we find a proper solution for this.

Another 'sticky' problem we have is that they're shown in every page results in  a custom Query Loop block that has `include sticky`.

https://user-images.githubusercontent.com/16275880/165917999-e374752f-a3b8-46e9-aad4-2acb746dd7b6.mov

**Review by ntsekouras — CHANGES_REQUESTED** [review]
I forgot to make the review more prominent not to merge right now because of the approval and [the breakage](https://github.com/WordPress/gutenberg/pull/40656#discussion_r861631990).

File: packages/block-library/src/post-template/index.php (hunk 1 above)
Comment (draganescu) [reply packages/block-library/src/post-template/index.php]: But it makes more sense, doesn't it to "merge the user settings into the defaults of the query" rather than the other way around? When you say:

> it breaks the whole query

can you instruct me how to, b/c testing seems to work as I expect it.

**Comment by draganescu** [comment]
@ntsekouras in https://github.com/WordPress/gutenberg/pull/40656/commits/066437cdfb68f9ebfef07dce588658163a4014f4  I added a change that unsets all query block defaults which are not editable by the user when inherit is `true`. That to me makes the most sense:

- when we use the `global WP_Query` use its defaults and settings
- allow stick post behavior to be specified even when using `global WP_Query` because sometimes we need sticky post exclusive query blocks: for example if I want to show sticky posts at the top in an archive page `global WP_Query` won't do it by default, but I can use a secondary query block.

**Comment by ntsekouras** [comment]
Haven't yet looked into the `sticky` use case here, but sharing a [comment of mine ](https://github.com/WordPress/gutenberg/pull/41348#issuecomment-1160486717)to a relevant PR about `inherit`.

TLDR:
>I think the big issue we're having with Query Loop and `inherit` is that while some controls are hidden in the UI, their values are preserved and I observed that while testing with at least `orderby` - this value is not set in the global query. We need to have no arbitrary results from block `query` attributes and what we expect when we set `inherit` to `true`. 

> Maybe we should `inherit` all options from global query but maybe there are some other concerns.. @draganescu had mentioned some nuances in the `sticky` use cases that I'm also planning to look into more..

**Comment by ntsekouras** [comment]
I'm not sure yet about this and how to approach it. Right now I'm thinking that it should be better to just hide the `sticky` control when we `inherit`. It seems confusing to have these two(and probably other controls) that override the `inherit` - or we should communicate that better by altering the descriptions of the controls, etc..

For the specific use case now (`sticky`) I feel that we shouldn't honor every option and we should hide the control too. For example it doesn't make sense to have `sticky -> only` and `inherit`. This way it will just show the sticky posts...

A solution could be specific to [sticky behavior
](https://developer.wordpress.org/themes/functionality/sticky-posts/), which is: show them only on front page if we `inherit` the global query. So inside [this block](https://github.com/WordPress/gutenberg/pull/40656/files#diff-4657bc4ddd88ab4cf8b48ad35b2fcf177edd40040c5976e2a3f9b89b52dd2bc7R51), something like this right after the merge of args might do the trick?
```
if ( is_front_page() ) {
	unset( $query_args['s'] );
}
```

What do you think?

**Comment by draganescu** [comment]
I think sticky posts are important for certain workflows. I would proceed from the feedback gathered here, plus a chat with @jameskoster, plus the fact that we don't really have straight answers to:

- revert changes to the `query` block's rendering
- simply hide any UI that contains overriding options for `WP_Query` if inherit is ON (do so programatically in the edit component not only to the "sticky" override).
- come back to it with better approaches, maybe even make a "sticky" posts block as a variation of query or something.

**Comment by spacedmonkey** [comment]
Will this functionality be ported to 6.0.1 or 6.0.2 Seems like a big enough bug that worth including.

**Comment by draganescu** [comment]
I did something and I see some changes that should not be here. Will fix.

@spacedmonkey I don't see why not.

**Review by spacedmonkey — APPROVED** [review]
This looks good!

CC @SergeyBiryukov @adamziel  for consideration in 6.0.1

**Comment by SergeyBiryukov** [comment]
I'm not familiar enough with this code for a proper review, but it looks like this would need a package update to be included in core. I see there were [some concerns](https://github.com/WordPress/gutenberg/pull/40656#discussion_r861631990) about some of the changes. I'll defer to the Gutenberg team on whether this should be included in 6.0.1 or 6.0.2 🙂

**Review by ntsekouras — COMMENTED** [review]
Thanks for iterating on this one @draganescu !

In general after testing this seems to work well. We would need to also make similar changes to [No Results block](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/query-no-results/index.php#L27).

A small concern would be this removed snippet:

> if ( empty( $query_args['post_type'] ) && is_singular() ) {
            $query_args['post_type'] = get_post_type( get_the_ID() );
}

But I don't think it breaks anything. My understanding is that this would just have some fallback in the case of `inherit` in a single page, where now it shows nothing. This is not a blocker for me as in general it doesn't make sense to have such a query in a single page. I think this is part of the context detection issue we have, where we should better handle the Query Loop based on the context it lives. --cc @aristath 

Finally, I don't think we should rush it in core, as it's not a bug that creates problems. To the contrary, it would be better to give it time to test it in the plugin for any regression/issues.

**Comment by aristath** [comment]
> My understanding is that this would just have some fallback in the case of inherit in a single page, where now it shows nothing. This is not a blocker for me as in general it doesn't make sense to have such a query in a single page. I think this is part of the context detection issue we have, where we should better handle the Query Loop based on the context it lives. --cc @aristath

If a page/post template does not exist, then the WP template system and hierarchies fallback to using the main `index` template. If I'm not mistaken that's why this was added... It was just an extra line of defense to accommodate scenarios like that.
It _is_ possible to have a block theme with nothing but an `index.html` template... and WP _should_ be able to handle these cases. So I should be able to build a theme with an `index.html` template which can handle both archives and single posts

**Comment by ntsekouras** [comment]
Thanks for explaining Ari!

> It is possible to have a block theme with nothing but an index.html template... and WP should be able to handle these cases. So I should be able to build a theme with an index.html template which can handle both archives and single posts

I just tested that(only `index.html` and `inherit:true`) and seems to work well because it actually adds the `name` param. I'd love some more testing though to be sure 😄

**Comment by adamziel** [comment]
@ntsekouras said:

> Finally, I don’t think we should rush it in core, as it’s not a bug that creates problems. To the contrary, it would be better to give it time to test it in the plugin for any regression/issues.

And so I won't be rushing this into 6.0.1 RC1 that happens in a few hours. Instead, I added the backport label to flag this for 6.0.2 release squad.

**Comment by draganescu** [comment]
@ntsekouras I'll add the updates to the no results block. @huubi and @spacedmonkey I've added you both as co-authors since the optimisation to the query was not part of the original intention or code of the PR.

**Comment by draganescu** [comment]
@ntsekouras while adding the update to the no results block I realized I don't understand why do we need to query anything if we render a no results page, since we know there are no results? What was the intention there?

Later edit: 

It seems this block could get the information if there are no posts from context rather than re-querying just to decide if it is going to render itself.

File: packages/block-library/src/query-no-results/index.php [hunk 2]
```diff
@@ -21,22 +21,23 @@ function render_block_core_query_no_results( $attributes, $content, $block ) {
...
-		}
+		$query = $wp_query;
+	} else {
+		$query_args = build_query_vars_from_query_block( $block, $page );
+		$query      = new WP_Query( $query_args );
 	}
-	$query = new WP_Query( $query_args );
 
-	if ( $query->have_posts() ) {
+	if ( ! $query->have_posts() ) {
```
Comment (ntsekouras) [inline packages/block-library/src/query-no-results/index.php]: We shouldn't change this logic. This block is actually a block of Query Loop and a way to show specific content/blocks if the Query doesn't have any results. So the previous check is correct to render nothing if the used query has results. Makes sense?

Comment (draganescu) [reply packages/block-library/src/query-no-results/index.php]: Ah this is a mistake from porting the changes from the query block, not intentional. So we render this block even if there are results. Got it.

Comment (draganescu) [reply packages/block-library/src/query-no-results/index.php]:
```suggestion
	if ( $query->have_posts() ) {
```

**Review by ntsekouras — APPROVED** [review]
Thanks @draganescu and also @huubi and @spacedmonkey for contributing in this one!

I'd appreciate some more testing about this https://github.com/WordPress/gutenberg/pull/40656#issuecomment-1174774280, but other than that looks good. With 🟢 CI let's ship.

**Review by aristath — APPROVED** [review]

File: packages/block-library/src/query-no-results/index.php [hunk 3]
```diff
@@ -21,22 +21,24 @@ function render_block_core_query_no_results( $attributes, $content, $block ) {
 
 	$page_key   = isset( $block->context['queryId'] ) ? 'query-' . $block->context['queryId'] . '-page' : 'query-page';
 	$page       = empty( $_GET[ $page_key ] ) ? 1 : (int) $_GET[ $page_key ];
```
Comment (aristath) [inline packages/block-library/src/query-no-results/index.php]: Just fixing CS alignments 👍 
```suggestion
	$page_key = isset( $block->context['queryId'] ) ? 'query-' . $block->context['queryId'] . '-page' : 'query-page';
	$page     = empty( $_GET[ $page_key ] ) ? 1 : (int) $_GET[ $page_key ];
```

**Comment by gziolo** [comment]
Cherry-picked with 493d14fffc89d3f124b160995559fb5c75b03a03 for WordPress 6.0.2 release.
